kqueen/auth/ldap.py

```diff
@@ -15,38 +16,50 @@ def __init__(self, *args, **kwargs):
         """
 
         super(LDAPAuth, self).__init__(*args, **kwargs)
+        if not all(hasattr(self, attr) for attr in ['uri', 'admin_dn', 'password']):
+            logger.error('Failed to configure LDAP synchronization error.')
+            raise ImproperlyConfigured
+
+        # Define Kqueen rdn for all dc's
+        dc_list = []
+        for i in ldap.dn.explode_dn(self.admin_dn):
+            if i.startswith('dc='):
+                dc_list.append(i)
+        self.kqueen_dc = ','.join(dc_list)
+
+        # Bind connection for Kqueen Read-only user
+        if self._bind(self.admin_dn, self.password):
+            self.connection = ldap.initialize(self.uri)
+            self.connection.simple_bind_s(self.admin_dn, self.password)
+            self.connection.protocol_version = ldap.VERSION3
+        else:
+            logger.error('Failed to bind connection for Kqueen Read-only user')
+            self.connection.unbind()
 
-        if not hasattr(self, 'uri'):
-            raise Exception('Parameter uri is required')
-
-        self.connection = ldap.initialize(self.uri)
-
-    @staticmethod
-    def _email_to_dn(email):
-        """This function reads email and converts it to LDAP dn
+    def _get_matched_dn(self, cn):
+        """This function reads username as cn and returns all matched full-dn's
 
         Args:
-            email (str): e-mail address
+            cn (str): Username of invited user
 
         Returns:
-            dn (str): LDAP dn, like 'cn=admin,dc=example,dc=org'
+            matched_dn (list): List of all matched dn's in groups.
         """
 
-        segments = []
-        if '@' in email:
-            cn, dcs = email.split('@')
-        else:
-            cn = email
-            dcs = ''
-
-        if cn:
-            segments.append('cn={}'.format(cn))
+        base_dn = self.kqueen_dc
+        search_scope = ldap.SCOPE_SUBTREE
+        retrieveAttributes = ['dn']
+        search_filter = "(cn={})".format(cn)
+        search_result = self.connection.search_s(base_dn, search_scope, search_filter, retrieveAttributes)
 
-        if '.' in dcs:
-            for s in dcs.split('.'):
-                segments.append('dc={}'.format(s))
+        if search_result:
```

Comment:
You don't need to test if `search_result` is empty, the generator will simply end up being empty list if it is. If you do it this way, you are again facing the issue, that `matched_dn` may not be defined when you want to return it, so it will raise error